File: src/com/mcblox/parkour/objects/Course.java
package com.mcblox.parkour.objects;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.block.data.BlockData;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

import com.mcblox.parkour.enums.FolderStructure;
import com.mcblox.parkour.utils.Console;

public class Course {

	// STATIC

	public static final int VIEW_DISTANCE = 3;
	
	public static List<Course> courses = new ArrayList<Course>();

	private static void resetIds() {
		for (int i = 0; i < courses.size(); i++) {
			courses.get(i).id = i;
		}
	}

	public static Course createCourse(String name) {
		Course output = new Course(name);
		courses.add(output);
		return output;
	}

	public static void deleteCourse(Course course) {
		courses.remove(course);
		resetIds();
		File file = new File(FolderStructure.PARKOUR_COURSES.file(), course.getName() + ".dat");
		file.delete();
		course.show();
		course = null;
	}

	public static void loadCourses() {
		for (File file : FolderStructure.PARKOUR_COURSES.file().listFiles()) {
			YamlConfiguration yaml = YamlConfiguration.loadConfiguration(file);

			// Initialize Course Load
			Course course = new Course();

			// Load initial variables
			course.rename(yaml.getString("name"));
			course.setSpawn((Location) yaml.get("spawn"));
			course.setStartBlock(((Location) yaml.get("start_block")).getBlock());
			course.setFinishBlock(((Location) yaml.get("finish_block")).getBlock());

			// Go into regions
			ConfigurationSection regions = yaml.getConfigurationSection("regions");
			for (String region_id : regions.getKeys(false)) {

				// Initialize region
				CourseRegion region = new CourseRegion(course);

				// Go into region blocks
				ConfigurationSection blocks = regions.getConfigurationSection(region_id);
				for (String block_id : blocks.getKeys(false)) {

					// Load block data
					Location loc = ((Location) blocks.get(block_id + ".loc"));
					BlockData blockdata = Bukkit.createBlockData(blocks.getString(block_id + ".data"));
					List<String> lines = new ArrayList<String>();
					lines = blocks.getStringList(block_id + ".lines");

					// Initialize CourseBlock
					CourseBlock block = new CourseBlock(loc, blockdata, lines);

					// Register course block into region
					region.addBlock(block);

					// -- End: for #Block Data
				}

				// Finish region load, and register region to course
				region.finishLoad();
				course.addRegion(region);

				// -- End: for #Region Blocks
			}

			// Register course into memory
			courses.add(course);

			// Hide Course
			course.hide();

			// -- End: for #Course Regions
		}

		// -- End: loadCourses()
	}

	// PUBLIC

	private int id;
	private String name;
	private Location spawn;
	private Block start;
	private Block finish;

	private List<CourseRegion> regions = new ArrayList<CourseRegion>();
	private HashMap<Player, Integer> playing = new HashMap<Player, Integer>();
	private List<Player> inQueue = new ArrayList<Player>();

	public Course(String name) {
		id = courses.size();
		this.name = name;
	}

	public Course() {
		id = courses.size();
	}

	public int getId() {
		return id;
	}

	public void rename(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	public void setSpawn(Location loc) {
		spawn = loc;
	}
	
	public Location getSpawn() {
		return spawn;
	}
	
	public void setStartBlock(Block start) {
		this.start = start;
	}

	public Block getStartBlock() {
		return start;
	}

	public void setFinishBlock(Block finish) {
		this.finish = finish;
	}

	public Block getFinishBlock() {
		return finish;
	}

	public void resetRegions() {
		regions.clear();
	}

	public void addRegion(CourseRegion region) {
		regions.add(region);
		for (int i = 0; i < regions.size(); i++) {
			region.id = i;
		}
	}

	public List<CourseRegion> getRegions() {
		return regions;
	}

	public boolean hasBlock(Block block) {
		for (CourseRegion region : regions) {
			if (region.hasBlock(block)) {
				return true;
			}
		}
		return false;
	}

	public void addQueue(Player player) {
		inQueue.add(player);
		updateVisibility();
	}

	public boolean isQueue(Player player) {
		return inQueue.contains(player);
	}

	public boolean isPlaying(Player player) {
		if (inQueue.contains(player)) {
			return true;
		}
		if (playing.containsKey(player)) {
			return true;
		}
		return false;
	}

	public void startPlayer(Player player) {
		inQueue.remove(player);
		playing.put(player, 0);
		updateVisibility();
	}

	public void removePlayer(Player player) {
		if (inQueue.contains(player)) {
			inQueue.remove(player);
		}
		if (playing.containsKey(player)) {
			playing.remove(player);
		}
		updateVisibility();
	}

	public Integer getPlayerPos(Player player) {
		return playing.get(player);
	}

	public void updatePlayerPos(Player player, Integer position) {
		playing.replace(player, position);
		updateVisibility();
	}

	public void show() {
		for (CourseRegion region : regions) {
			region.show();
		}
	}

	public void hide() {
		for (CourseRegion region : regions) {
			region.hide();
		}
	}

	public void updateVisibility() {

		// Initialize variables
		List<Integer> positionsPlaying = new ArrayList<Integer>();
		HashMap<Integer, Boolean> visiblePlan = new HashMap<Integer, Boolean>();

		// Setup visibility plan
		for (CourseRegion region : regions) {
			visiblePlan.put(region.id, false);
		}

		// Check if someone is in queue
		if (inQueue.size() != 0) {
			visiblePlan.replace(0, true);
		}

		// Add the region positions that are currently occupying players
		playing.forEach((k, v) -> {
			if (!positionsPlaying.contains(v)) {
				positionsPlaying.add(v);
			}
		});

		// Show regions that are in front of the regions occupied by the players
		for (Integer pos : positionsPlaying) {
			for (int i = 0; i < VIEW_DISTANCE; i++) {
				visiblePlan.replace(pos + i, true);
			}
		}

		// Execute visibility plan
		visiblePlan.forEach((k, v) -> {
			if (regions.get(k).isVisible()) {
				if (!v) {
					regions.get(k).hide();
				}
			}
			else {
				if (v) {
					regions.get(k).show();
				}
			}
		});

		// -- End: updateVisibility()
	}

	public void saveCourse() {
		File file = new File(FolderStructure.PARKOUR_COURSES.file(), name + ".dat");

		try {

			// Initialize writer buffer
			YamlConfiguration yaml = new YamlConfiguration();

			// Save initial variables
			yaml.set("name", name);
			yaml.set("spawn", spawn);
			yaml.set("start_block", start.getLocation());
			yaml.set("finish_block", finish.getLocation());

			// Save regions
			for (CourseRegion region : regions) {

				// Save region blocks
				for (int i = 0; i < region.getBlocks().size(); i++) {

					// Serialize region blocks
					CourseBlock block = region.getBlocks().get(i);
					yaml.set("regions." + region.id + "." + i + ".loc", block.getLocation());
					yaml.set("regions." + region.id + "." + i + ".data", block.getBlockData().getAsString());
					yaml.set("regions." + region.id + "." + i + ".lines", block.getLines());

				}

			}

			// End save
			yaml.save(file);

			// -- End: try_catch
		} catch (IOException e) {
			Console.catchError(e, "Course.saveCourse()");
		}

		// -- End: saveCourse()
	}

}

File: src/org/serversmc/parkour/objects/Course.kt
package org.serversmc.parkour.objects

import org.bukkit.*
import org.bukkit.configuration.file.*
import org.bukkit.entity.*
import org.serversmc.parkour.interfaces.*
import org.serversmc.parkour.utils.*
import org.serversmc.parkour.utils.Console
import java.io.*
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.math.*
class Course(private val file: File) {
	
	enum class Mode {
		OPEN, CLOSED
	}
	
	private val viewDistance = 3
	
	private val players = HashMap<Player, PlayerData>()
	private val regions = ArrayList<CRegion>()
	private val sensors = ArrayList<CSensor>()
	
	private var plays = 0
	private var wins = 0
	
	private lateinit var author: UUID
	private val spawn: CSpawn = CSpawn()
	private var name = file.nameWithoutExtension
	private var mode = Mode.CLOSED
	
	fun load() {
		// Check if file exists
		if (file.exists()) {
			// Load yaml file
			val yaml = YamlConfiguration.loadConfiguration(file)
			// Initialize attributes
			name = yaml.getString("name")!!
			spawn.setLocation(yaml.get("spawn") as? Location)
			mode = Mode.valueOf(yaml.getString("mode")!!)
			author = UUID.fromString(yaml.getString("author")!!)
			// Try to load regions
			val yamlRegions = yaml.getConfigurationSection("regions")
			yamlRegions?.getKeys(false)?.forEach {
				// Check if section is valid
				if (it == null) return@forEach
				val section = yamlRegions.getConfigurationSection(it) ?: return@forEach
				// Load region
				regions.add(CRegion().apply { load(section) })
			}
			// Try to load sensors
			val yamlSensors = yaml.getConfigurationSection("sensors")
			yamlSensors?.getKeys(false)?.forEach {
				// Check if section is valid
				if (it == null) return@forEach
				val section = yamlSensors.getConfigurationSection(it) ?: return@forEach
				// Load sensor
				sensors.add(CSensor().apply { load(section) })
			}
		}
		else {
			// Create new data file
			file.createNewFile()
		}
		// Validate if course isReady
		if (mode == Mode.OPEN) {
			if (!isReady()) {
				Console.warn("Course: \"$name\" loaded as open, but does not qualify as ready.")
				mode = Mode.CLOSED
			}
		}
		// Show / Hide depending on mode
		if (mode == Mode.OPEN) hide()
		else show()
	}
	
	/*************/
	/** PLAYERS **/
	/*************/
	
	fun getPlayers(): ArrayList<Player> {
		return ArrayList<Player>().apply {
			players.forEach { (t, _) ->
				add(t)
			}
		}
	}
	
	fun addPlayer(player: Player) {
		players[player] = PlayerData()
		updateVisibility()
	}
	
	fun removePlayer(player: Player) {
		players.remove(player)
		updateVisibility()
	}
	
	fun hasPlayer(player: Player) = players.containsKey(player)
	
	fun getPlayerPos(player: Player) = players[player]!!.getPosition()
	
	fun setPlayerPos(player: Player, id: Int) {
		players[player]!!.setPosition(id)
		updateVisibility()
	}
	
	fun getPlayerCheckpoint(player: Player) = players[player]!!.getCheckpoint()
	
	fun setPlayerCheckpoint(player: Player, sensor: CSensor) {
		players[player]!!.setCheckpoint(sensor)
	}
	
	/*************/
	/** REGIONS **/
	/*************/
	
	fun getRegions() = regions
	
	fun getRegion(index: Int) = regions.getOrNull(index)
	
	fun createRegion(world: World): CRegion {
		return CRegion().apply {
			setId(regions.size)
			setWorld(world)
			regions.add(this)
		}
	}
	
	fun removeRegion(region: CRegion) {
		regions.remove(region)
		region.hideHologram()
		regions.forEachIndexed { index, cRegion ->
			cRegion.setId(index)
			cRegion.updateHoloName()
		}
	}
	
	/*************/
	/** SENSORS **/
	/*************/
	
	fun getSensors() = sensors
	
	fun hasSensor(location: Location): Boolean {
		sensors.forEach { if (it.getLocation() == location) return true }
		return false
	}
	
	fun getSensor(location: Location) = sensors.singleOrNull { it.getLocation() == location }
	
	fun getStartSensor() = sensors.singleOrNull { it.getType() == CSensor.Type.START }
	
	fun setStartSensor(location: Location?) {
		sensors.singleOrNull { it.getType() == CSensor.Type.START }?.apply {
			this.hideHologram()
			sensors.remove(this)
		}
		if (location == null) return
		sensors.add(CSensor.create(CSensor.Type.START, location).apply { showHologram() })
	}
	
	fun getCheckpoints() = sensors.filter { it.getType() == CSensor.Type.CHECKPOINT }
	
	fun addCheckpoint(location: Location) {
		sensors.add(CSensor.create(CSensor.Type.CHECKPOINT, location).apply { showHologram() })
	}
	
	fun getFinishSensor() = sensors.singleOrNull { it.getType() == CSensor.Type.FINISH }
	
	fun setFinishSensor(location: Location?) {
		sensors.singleOrNull { it.getType() == CSensor.Type.FINISH }?.apply {
			hideHologram()
			sensors.remove(this)
		}
		if (location == null) return
		sensors.add(CSensor.create(CSensor.Type.FINISH, location).apply { showHologram() })
	}
	
	/***********************/
	/** GETTERS / SETTERS **/
	/***********************/
	
	fun getId() = CourseManager.getCourses().indexOf(this)
	
	fun getViewDistance() = viewDistance
	
	fun getSpawn() = spawn.getLocation()!!
	
	fun setSpawn(location: Location) {
		val tempLocation = location.clone().apply {
			x = location.blockX + 0.5
			y = location.blockY + 0.5
			z = location.blockZ + 0.5
			pitch = ((location.pitch / 15).roundToInt() * 15f)
			yaw = (location.yaw / 15).roundToInt() * 15f
		}
		spawn.setLocation(tempLocation.clone())
		updateHolograms()
	}
	
	fun getName() = name
	
	fun setName(name: String) {
		this.name = name
	}
	
	fun getMode() = mode
	
	fun setOpen() {
		mode = Mode.OPEN
		SelectManager.get(this).forEach {
			SelectManager.remove(it)
		}
		hide()
	}
	
	fun isOpen() = (mode == Mode.OPEN)
	
	fun setClosed() {
		mode = Mode.CLOSED
		players.forEach {
			it.key.teleport(getSpawn())
		}
		players.clear()
		show()
	}
	
	fun getAuthor() = author
	
	fun setAuthor(player: Player) {
		author = player.uniqueId
	}
	
	fun getPlays() = plays
	
	fun getWins() = wins
	
	fun getAveragePlayTime(): Long {
		// TODO - Return average time it takes to complete
		return 0
	}
	
	fun isReady(): Boolean {
		return when {
			(spawn.getLocation() == null) -> false
			(getStartSensor() == null) -> false
			(getFinishSensor() == null) -> false
			(regions.isEmpty()) -> false
			else -> true
		}
	}
	
	/***************/
	/** MODIFIERS **/
	/***************/
	
	private fun showHolograms() {
		sensors.forEach { (it as IHologram).showHologram() }
		regions.forEach { (it as IHologram).showHologram() }
		spawn.showHologram()
	}
	
	private fun hideHolograms() {
		sensors.forEach { (it as IHologram).hideHologram() }
		regions.forEach { (it as IHologram).hideHologram() }
		spawn.hideHologram()
	}
	
	fun updateHolograms() {
		if (isOpen()) {
			hideHolograms()
			return
		}
		if (SelectManager.contains(this)) {
			showHolograms()
		}
		else {
			hideHolograms()
		}
	}
	
	private fun updateVisibility() {
		val show = ArrayList<Int>()
		players.values.forEach {
			for (i in 0..viewDistance) show.add(it.getPosition() + i)
		}
		regions.forEach {
			if (show.contains(it.getId())) it.show()
			else it.hide()
		}
	}
	
	fun save() {
		// Initialize variables
		val yaml = YamlConfiguration()
		// Add course attributes
		yaml.set("name", name)
		yaml.set("spawn", spawn.getLocation())
		yaml.set("mode", mode.name)
		yaml.set("author", author.toString())
		// Save Sensors
		sensors.forEachIndexed { id, sensor ->
			yaml.set("sensors.$id.type", sensor.getType().name)
			yaml.set("sensors.$id.loc", sensor.getLocation())
		}
		// Save Regions
		regions.forEachIndexed { regionId, region ->
			yaml.set("regions.$regionId.world", region.getWorld().uid.toString())
			region.getBlocks().forEachIndexed { blockId, b ->
				yaml.set("regions.$regionId.$blockId.loc", b.getLocation())
				yaml.set("regions.$regionId.$blockId.data", b.getBlockData().getAsString(true))
			}
		}
		// Save file
		yaml.save(file)
		hideHolograms()
	}
	
	fun delete() {
		file.delete()
		hideHolograms()
	}
	
	fun show() {
		regions.forEach { it.show() }
	}
	
	fun hide() {
		regions.forEach { it.hide() }
	}
	
}

